internal/proxy/role_handler.go

```diff
@@ -0,0 +1,270 @@
+	})
+	if err != nil {
+		err = fmt.Errorf("creating role %s: %w", body, err)
+		handleJSONErrorResponse(th.log, w, http.StatusInternalServerError, err)
+		return err
+	}
+	w.WriteHeader(http.StatusCreated)
+	return nil
+}
+
+func (th *RoleHandler) updateHandler(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	span := trace.SpanFromContext(ctx)
+
+	// read request body
+	var body CreateRoleBody
+	err := json.NewDecoder(r.Body).Decode(&body)
+	if err != nil {
+		err = fmt.Errorf("decoding request body: %w", err)
+		handleJSONErrorResponse(th.log, w, http.StatusBadRequest, err)
+		return err
+	}
+
+	setAttributes(span, map[string]interface{}{
+		"name":        body.Name,
+		"storageType": body.StorageType,
+		"systemId":    body.SystemID,
+		"pool":        body.Pool,
+		"quota":       body.Quota,
+	})
+	th.log.WithFields(logrus.Fields{
+		"name":        body.Name,
+		"storageType": body.StorageType,
+		"systemId":    body.SystemID,
+		"pool":        body.Pool,
+		"quota":       body.Quota,
+	}).Info("Requesting role update")
+
+	_, err = th.client.Update(ctx, &pb.RoleUpdateRequest{
+		Name:        body.Name,
+		StorageType: body.StorageType,
+		SystemId:    body.SystemID,
+		Pool:        body.Pool,
+		Quota:       body.Quota,
+	})
+
+	if err != nil {
+		err = fmt.Errorf("updating role %s: %w", body, err)
+		handleJSONErrorResponse(th.log, w, http.StatusInternalServerError, err)
+		return err
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+	return nil
+}
+
+func (th *RoleHandler) getHandler(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	span := trace.SpanFromContext(ctx)
+
+	// parse role name from request parameters
+	params := r.URL.Query()["name"]
+
+	// not querying one role but list all
+	if len(params) == 0 || params[0] == "" {
+		th.log.Info("Requesting role list")
+
+		// call role service
+		roles, err := th.client.List(ctx, &pb.RoleListRequest{})
+		if err != nil {
+			err = fmt.Errorf("listing roles: %w", err)
+			handleJSONErrorResponse(th.log, w, http.StatusInternalServerError, err)
+			return err
+		}
+
+		// write roles to client
+		err = json.NewEncoder(w).Encode(&roles)
+		if err != nil {
+			err = fmt.Errorf("writing role list response: %w", err)
+			handleJSONErrorResponse(th.log, w, http.StatusInternalServerError, err)
+			return err
+		}
+		return nil
+	}
+	// else, call role service to get one specific role
+	name := params[0]
+
+	setAttributes(span, map[string]interface{}{
+		"name": name,
+	})
+	th.log.WithFields(logrus.Fields{
+		"name": name,
+	}).Info("Requesting role get")
+
+	// call role service
+	role, err := th.client.Get(ctx, &pb.RoleGetRequest{
+		Name: name,
+	})
+	if err != nil {
+		err = fmt.Errorf("getting role %s: %w", name, err)
+		handleJSONErrorResponse(th.log, w, http.StatusInternalServerError, err)
+		return err
+	}
+
+	// return role to client
+	_, err = fmt.Fprint(w, protojson.MarshalOptions{Multiline: true, EmitUnpopulated: true, Indent: ""}.Format(role))
```

Comment:
You can use the standard json encoding here. This was necessary for getting a tenant to handle empty fields. It shouldn't be necessary for writing the role.

Reply:
I'll change it to `protojson.Format(role)`. storage_handler is using the same MarchalOptions I guess all from the Tenant template.

Reply:
fixed